### PR TITLE
WIP(router): Add useBlocker hook

### PR DESCRIPTION
**Pull request by xmaxcooking**
New version of  #9820 based on @Tobbe 's [input](https://github.com/redwoodjs/redwood/pull/9820#issuecomment-1890488791)

Conceptually it would be used like this:
```ts
const PostForm = (props: PostFormProps) => {
  const form = useForm<FormPost>()
  const blocker = useBlocker(form.formState.isDirty)

  return (
    <div className="rw-form-wrapper">
      <Form formMethods={form} onSubmit={onSubmit} error={props.error}>
        State {blocker.state}
        {blocker.state === 'BLOCKED' ? (
          <div>
            <button type="button" onClick={() => blocker.confirm()}>
              Confirm
            </button>
            <button type="button" onClick={() => blocker.abort()}>
              Abort
            </button>
          </div>
        ) : null}
       ...
      </Form>
    </div>
   )
}
```

### Help wanted

I'm currently stuck with the history.replaceState logic and I'm opening this PR just to see if someone has an idea what's happening here. Feel free to comment if you have a suggestion.
I'd be happy to try it out.

### Behaviour

The new navigation provider is patching the global history object to include the custom checks. If any blocker is `true` it shouldn't invoke the pushState and keep the current location but it somehow desyncs the active route from the URL. 
The blocker prevents the new page from being loaded but the window.location changes and on the next navigation it doesn't block anymore because it doesn't know it's on a page that should be blocked.

```ts
globalThis.history.pushState = function pushState(
  data: any,
  unused: string,
  url?: string | URL | null
) {
  const currentLocation = new URL(globalThis.location.href)
  const newLocation = new URL(globalThis.location.origin + url)

  if (
    currentLocation.pathname === newLocation.pathname &&
    currentLocation.search === newLocation.search &&
    currentLocation.hash === newLocation.hash
  ) {
    return
  }

  const blocked = blockers.some(({ check }) =>
    check(currentLocation, newLocation, NavigationMethod.PUSH, () => {
      originalPushState(data, unused, url)
    })
  )

  if (!blocked) {
    originalPushState(data, unused, url)
  }
}
```

### CODE REVIEW

File: packages/router/src/history.tsx [hunk 1]
```diff
@@ -15,20 +15,10 @@ const createHistory = () => {
...
-      if (
-        globalThis?.location?.pathname !== pathname ||
-        globalThis?.location?.search !== search ||
-        globalThis?.location?.hash !== hash
-      ) {
-        if (options?.replace) {
-          globalThis.history.replaceState({}, '', to)
-        } else {
-          globalThis.history.pushState({}, '', to)
-        }
```
Comment (xmaxcooking) [inline packages/router/src/history.tsx:31]: The same location check is obsolete with the patch of globalThis.history.replaceState and globalThis.history.pushState

File: packages/router/src/navigation.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  })
+
+  useEffect(() => {
+    setNavigation((prev) => {
+      return {
+        from: prev.to,
+        to: location,
+      }
+    })
+  }, [location])
```
Comment (xmaxcooking) [inline packages/router/src/navigation.tsx:78]: navigation is used to check blocking of unload events.

File: packages/router/src/navigation.tsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        currentLocation.search === newLocation.search &&
+        currentLocation.hash === newLocation.hash
+      ) {
+        return
+      }
+
+      const blocked = blockers.some(({ check }) =>
+        check(currentLocation, newLocation, NavigationMethod.PUSH, () => {
+          originalPushState(data, unused, url)
+        })
```
Comment (xmaxcooking) [inline packages/router/src/navigation.tsx:110]: pass the current and new location to the check function and a callback to invoke the navigation when the user confirms.

File: packages/router/src/navigation.tsx [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+      )
+
+      if (!blocked) {
+        originalPushState(data, unused, url)
+      }
```
Comment (xmaxcooking) [inline packages/router/src/navigation.tsx:115]: pushState isn't invoke when a blocker returned true... but the URL changes to the new location anyway. (WHY??)

File: packages/router/src/router.tsx [hunk 5]
```diff
@@ -73,13 +74,15 @@ const Router: React.FC<RouterProps> = ({
...
     // Wrap it in the provider so that useLocation can be used
     <LocationProvider trailingSlashes={trailingSlashes}>
-      <LocationAwareRouter
-        useAuth={useAuth}
-        paramTypes={paramTypes}
-        pageLoadingDelay={pageLoadingDelay}
-      >
-        {children}
-      </LocationAwareRouter>
+      <NavigationProvider>
```
Comment (xmaxcooking) [inline packages/router/src/router.tsx:77]: could also be named HistoryProvider

File: packages/router/src/useBlocker.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  useEffect(() => {
+    window.addEventListener('beforeunload', unloadBlocker)
+    const blockerId = register(navigationBlocker)
+    return () => {
+      window.removeEventListener('beforeunload', unloadBlocker)
+      unregister(blockerId)
+    }
+  }, [register, unregister, unloadBlocker, navigationBlocker])
+
+  const confirm = useCallback(() => {
```
Comment (xmaxcooking) [inline packages/router/src/useBlocker.ts:92]: this is never called. WHY?

**Review by xmaxcooking — COMMENTED** [review]
Comment

**Comment by xmaxcooking** [comment]
For anyone wondering about this PR. I haven't been able to get back to this but it's still on my list.

**Comment by thedavidprice** [comment]
Lots of work on this @xmaxcooking Thank you! Checking in to see if this is something you'd like us to keep open — fine on our end.

**Comment by xmaxcooking** [comment]
> Lots of work on this @xmaxcooking Thank you! Checking in to see if this is something you'd like us to keep open — fine on our end.

feel free to close it. the discussion with @Tobbe helped bring this into the right direction but at this point I'll open it again when I have a working state.

**Comment by thedavidprice** [comment]
Understood. We'd welcome you opening it up again in the future!

**Pull request closed without merging**